### PR TITLE
vimPlugins.cord-nvim: init at 2024-07-19

### PR DESCRIPTION
**Pull request by redyf**
## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [x] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by GaetanLepage** [comment]
Result of `nixpkgs-review pr 319576` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>vimPlugins.cord-nvim</li>
  </ul>
</details>

**Comment by GaetanLepage** [comment]
Result of `nixpkgs-review pr 319576` run on aarch64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>vimPlugins.cord-nvim</li>
  </ul>
</details>

**Comment by GaetanLepage** [comment]
Result of `nixpkgs-review pr 319576` run on aarch64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>vimPlugins.cord-nvim</li>
  </ul>
</details>

**Comment by GaetanLepage** [comment]
Result of `nixpkgs-review pr 319576` run on x86_64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>vimPlugins.cord-nvim</li>
  </ul>
</details>

**Comment by GaetanLepage** [comment]
This will probably won't work as is as the plugin is partly written in rust and this needs to be compiled.
Have a look at the `sniprun`, `codesnap-nvim` or `nvim-spectre` examples for inspiration (in `overrides.nix`).

**Comment by redyf** [comment]
I'll take a look, thank you very much 🙏

**Comment by redyf** [comment]
@GaetanLepage I tried doing as you suggested and the package was built successfully, if you don't mind can you take a look at the package to make sure I'm in the right direction? Also thanks for telling me about those plugins, they helped a lot :smile:

**Review by GaetanLepage — COMMENTED** [review]
This looks quite good.
However, you now exhaustively define the plugin in `overrides.nix`.
Hence, you can completely get rid of the snippets in `vim-plugin-names` and `generated.nix`.
Also, wasn't it possible to use `buildRustPackage` in a `let`-`in` block for the rust dependency ?

**Comment by redyf** [comment]
> This looks quite good. However, you now exhaustively define the plugin in `overrides.nix`. Hence, you can completely get rid of the snippets in `vim-plugin-names` and `generated.nix`. Also, wasn't it possible to use `buildRustPackage` in a `let`-`in` block for the rust dependency ?

Indeed, it makes more sense. The changes have been applied :+1:

File: pkgs/applications/editors/vim/plugins/overrides.nix [hunk 1]
```diff
@@ -545,6 +545,59 @@
     patches = [ ./patches/coq_nvim/emulate-venv.patch ];
   };
 
+  cord-nvim =
+    let
+      version = "2024-07-19";
+      src = fetchFromGitHub {
+      owner = "vyfor";
```
Comment (GaetanLepage) [inline pkgs/applications/editors/vim/plugins/overrides.nix:552]: I think that the indentation of the entire snippet is off.

Comment (redyf) [reply pkgs/applications/editors/vim/plugins/overrides.nix]: Should be good now, I formatted the snippet.

File: pkgs/applications/editors/vim/plugins/overrides.nix [hunk 2]
```diff
@@ -545,6 +545,58 @@
...
+      };
+      rustPackage = rustPlatform.buildRustPackage {
+        pname = "cord.nvim-rust";
+        inherit version src;
+
+        cargoSha256 = "sha256-6FYf4pHEPxvhKHHPmkjQ40zPxaiypnpDxF8kNH+h+tg=";
+
+        buildPhase = ''
+          cargo build --release --lib
+        '';
```
Comment (GaetanLepage) [inline pkgs/applications/editors/vim/plugins/overrides.nix:565]: You have manually overriden the `buildPhase` to add `--lib` ?
If so, you might be able to simply pass it using `cargoBuildFlags` ?

File: pkgs/applications/editors/vim/plugins/overrides.nix [hunk 3]
```diff
@@ -545,6 +545,60 @@
...
+      buildPhase = ''
+        cp ${rustPackage}/lib/cord.so cord.so
+      '';
+
+      installPhase = let
+        extension = if stdenv.isDarwin then "dylib" else "so";
+      in ''
+        mkdir -p $out/lua
+        cp cord $out/lua/cord.${extension}
+      '';
```
Comment (eljamm) [inline pkgs/applications/editors/vim/plugins/overrides.nix:592]:
```suggestion
      buildPhase = ''
        cp ${rustPackage}/lib/cord.${extension} cord.${extension}
      '';

      installPhase = ''
        install -D cord $out/lua/cord.${extension}
      '';
```

File: pkgs/applications/editors/vim/plugins/overrides.nix (hunk 2 above)
Comment (redyf) [reply pkgs/applications/editors/vim/plugins/overrides.nix]: Done, I had no idea about cargoBuildFlags.

**Review by eljamm — CHANGES_REQUESTED** [review]

File: pkgs/applications/editors/vim/plugins/overrides.nix [hunk 4]
```diff
@@ -545,6 +545,53 @@
...
+    buildVimPlugin {
+      pname = "cord.nvim";
+      inherit version src;
+
+      nativeBuildInputs = [
+        rustPackage
+      ];
+
+      buildPhase = ''
+        cp ${rustPackage}/lib/cord.${extension} cord.${extension}
```
Comment (eljamm) [inline pkgs/applications/editors/vim/plugins/overrides.nix:580]: I think you forgot to change the commit description after the squish as it has all those `refactor` comments. Can you remove them?

While you're at it, can you also change this `cp` to `install` to make things more consistent?

```suggestion
        install -D ${rustPackage}/lib/cord.${extension} cord.${extension}
```

**Review by eljamm — APPROVED** [review]
Thanks, this looks good to me :+1:

File: pkgs/applications/editors/vim/plugins/overrides.nix (hunk 4 above)
Comment (redyf) [reply pkgs/applications/editors/vim/plugins/overrides.nix]: Oops, I thought it was good practice to keep the description like that. I just removed them and rebased. All changes have been applied. :+1:

**Review by GaetanLepage — APPROVED** [review]
LGTM ! Thanks for addressing all the requests.
Also, thank you @eljamm for your detailed review !

**Comment by GaetanLepage** [comment]
@ofborg build vimPlugins.cord-nvim

**Pull request merged**